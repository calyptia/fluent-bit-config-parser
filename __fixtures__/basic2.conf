[INPUT]
    name forward.0
    tag forward.0
    Buffer_Chunk_Size 1M
    Buffer_Max_Size 6M
[INPUT]
    name fluentbit_metrics.1
    tag _calyptia_cloud
    scrape_on_start true
    scrape_interval 30
[INPUT]
    name fluentbit_metrics.2
    tag _calyptia_cloud
    scrape_on_start true
    scrape_interval 30

[OUTPUT]
    name  stdout.0
    match *
    retry_limit 1

[OUTPUT]
    name  calyptia.1
    match _calyptia_cloud
    tls   on
    tls.verify     on
    retry_limit 1
    api_key __REDACTED__
    cloud_port 443
    cloud_host cloud-api-dev.calyptia.com
    cloud_bla 1
