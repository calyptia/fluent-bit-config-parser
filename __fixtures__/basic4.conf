[INPUT]
    Name tail
    Alias erlang_tail
    Path babysitter.log,debug.log,reports.log
    Multiline On
    Parser_Firstline couchbase_erlang_multiline
    Refresh_Interval 10
    Skip_Long_Lines On
    Skip_Empty_Lines On
    Path_Key filename
    Tag couchbase.log.<logname>
    Tag_Regex ${COUCHBASE_LOGS}/(?<logname>[^.]+).log$
