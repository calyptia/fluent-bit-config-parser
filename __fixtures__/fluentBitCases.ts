import fs from 'fs';
import { COMMANDS, FluentBitSchemaType } from '../src/constants';

type Case = [name: string, content: string, SchemaAST: { config: FluentBitSchemaType[] }];

export const cases: Case[] = [
  [
    './__fixtures__/basic.conf',
    fs.readFileSync('./__fixtures__/basic.conf', { encoding: 'utf8' }),
    {
      config: [
        {
          command: COMMANDS.INPUT,
          name: 'tail',
          optional: {
            tag: 'tail.01',
            path: '/var/log/system.log',
          },
          id: 'UNIQUE',
        },
        {
          command: COMMANDS.OUTPUT,
          name: 's3',
          optional: {
            match: '*',
            bucket: 'your-bucket',
            region: 'us-east-1',
            store_dir: '/home/ec2-user/buffer',
            total_file_size: '50M',
            upload_timeout: '10m',
          },
          id: 'UNIQUE',
        },
        {
          command: COMMANDS.OUTPUT,
          name: 'splunk',
          optional: {
            match: '*',
            host: '127.0.0.1',
            port: '8088',
            tls: 'On',
            'tls.verify': 'Off',
            message_key: 'my_key',
          },
          id: 'UNIQUE',
        },
      ],
    },
  ],
  [
    './__fixtures__/basic2.conf',
    fs.readFileSync('./__fixtures__/basic2.conf', { encoding: 'utf8' }),
    {
      config: [
        {
          command: COMMANDS.INPUT,
          name: 'forward.0',
          optional: {
            tag: 'forward.0',
            buffer_chunk_size: '1M',
            buffer_max_size: '6M',
          },
          id: 'UNIQUE',
        },
        {
          command: COMMANDS.OUTPUT,
          name: 'stdout.0',
          optional: {
            match: '*',
            retry_limit: '1',
          },
          id: 'UNIQUE',
        },
      ],
    },
  ],
  [
    './__fixtures__/basic3.conf',
    fs.readFileSync('./__fixtures__/basic3.conf', { encoding: 'utf8' }),
    {
      config: [
        {
          command: COMMANDS.INPUT,
          name: 'Tail',
          optional: {
            tag: 'kube.*',
            path: '/foo',
          },
          id: 'UNIQUE',
        },
        {
          command: COMMANDS.OUTPUT,
          name: 'Splunk',
          optional: {
            match: 'kube*audit',
            splunk_token: 'foo',
          },
          id: 'UNIQUE',
        },
        {
          command: COMMANDS.OUTPUT,
          name: 'kinesis_firehose',
          optional: {
            region: 'oregon',
            delivery_stream: 'foo',
            match: 'kube.*',
          },
          id: 'UNIQUE',
        },
        {
          command: COMMANDS.FILTER,
          name: 'kubernetes',
          optional: {
            match: 'kube*',
          },
          id: 'UNIQUE',
        },
      ],
    },
  ],
  [
    './__fixtures__/basic4.conf',
    fs.readFileSync('./__fixtures__/basic4.conf', { encoding: 'utf-8' }),
    {
      config: [
        {
          command: COMMANDS.INPUT,
          name: 'tail',
          optional: {
            alias: 'erlang_tail',
            path: 'babysitter.log,debug.log,reports.log',
            multiline: 'On',
            parser_firstline: 'couchbase_erlang_multiline',
            refresh_interval: '10',
            skip_long_lines: 'On',
            skip_empty_lines: 'On',
            path_key: 'filename',
            tag: 'couchbase.log.<logname>',
            tag_regex: '${COUCHBASE_LOGS}/(?<logname>[^.]+).log$',
          },
          id: 'UNIQUE',
        },
      ],
    },
  ],
];
