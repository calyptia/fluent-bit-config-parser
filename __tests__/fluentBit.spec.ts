import { FluentBitSchema } from '../index';
import { cases } from '../__fixtures__/fluentBitCases';

jest.mock('uuid', () => ({ v4: () => 'UNIQUE' }));

describe('fluentBit', () => {
  it('Fails if config is empty', () => {
    expect(() => new FluentBitSchema('       ')).toThrowErrorMatchingInlineSnapshot('"Invalid Config file"');
  });

  it('Fails if config has no fields', () => {
    expect(() => new FluentBitSchema('# some comment')).toThrowErrorMatchingInlineSnapshot(
      '"We could not find fields in the configuration"'
    );
  });
  it('Fails if config has invalid commands', () => {
    expect(
      () =>
        new FluentBitSchema(`
    [INVALID]
        Name        tail
        Tag         tail.01
        Path        /var/log/system.log

    [OUTPUT]
        Name        splunk
        Match       *
        Host        127.0.0.1
        Port        8088
        TLS         On
        TLS.Verify  Off
        Message_Key my_key
    `)
    ).toThrowErrorMatchingInlineSnapshot(
      '"Command is not valid, we got INVALID, it should be OUTPUT,INPUT,FILTER,SERVICE,PARSER"'
    );
  });

  it('Should ignore new line comments on AST', () => {
    const rawConfig = `
    [INPUT]
        # new line comment
        Name        tail
        Tag         tail.01
        Path        /var/log/system.log

    [OUTPUT]
        Name s3
        Match *
        bucket your-bucket
        region us-east-1
        store_dir /home/ec2-user/buffer
        total_file_size 50M
        upload_timeout 10m 
    `;
    const config = new FluentBitSchema(rawConfig);
    expect(config.schema).toMatchInlineSnapshot(`
      Object {
        "config": Array [
          Object {
            "command": "INPUT",
            "id": "UNIQUE",
            "name": "tail",
            "optional": Object {
              "path": "/var/log/system.log",
              "tag": "tail.01",
            },
          },
          Object {
            "command": "OUTPUT",
            "id": "UNIQUE",
            "name": "s3",
            "optional": Object {
              "bucket": "your-bucket",
              "match": "*",
              "region": "us-east-1",
              "store_dir": "/home/ec2-user/buffer",
              "total_file_size": "50M",
              "upload_timeout": "10m",
            },
          },
        ],
      }
    `);
  });
  it.each(cases)('Parse config: %s', (_name, rawConfig, expected) => {
    const config = new FluentBitSchema(rawConfig);
    expect(config.schema).toMatchObject(expected);
  });

  it.each(cases)('Returns source: %s', (_name, rawConfig) => {
    const config = new FluentBitSchema(rawConfig);
    expect(config.source).toBe(rawConfig);
  });
  it('should transform schema to string for basic.conf', () => {
    const [, rawConfig] = cases[0];

    // We need to normalize given that SchemaToString will return values toLowerCase + spaces normalized.
    // const normalize = (config: string) => config.replace(/\s/g, '').toLocaleLowerCase();

    const config = new FluentBitSchema(rawConfig);

    expect(config.toString()).toMatchInlineSnapshot(`
      "[INPUT]
          name  tail # some comment
          tag  tail.01
          path  /var/log/system.log

      [OUTPUT]
          name  s3
          match  *
          bucket  your-bucket
          region  us-east-1
          store_dir  /home/ec2-user/buffer
          total_file_size  50M
          upload_timeout  10m

      [OUTPUT]
          name  splunk
          match  *
          host  127.0.0.1
          port  8088
          tls  On
          tls.verify  Off
          message_key  my_key"
    `);
  });

  it.each(cases)('is %s, fluent-bit configuration?', (_name, rawConfig) => {
    expect(FluentBitSchema.isFluentBitConfiguration(rawConfig)).toBe(true);
  });

  it('fluentD.conf should not be fluent-bit configuration', () => {
    const fluentDConfig = `
    #  Receive events from 24224/tcp
    # This is used by log forwarding and the fluent-cat command
    <source>
      @type forward
      port 24224
    </source>
    
    # http://<ip>:9880/myapp.access?json={"event":"data"}
    <source>
      @type http
      port 9880
    </source>
    
    # Match events tagged with "myapp.access" and
    # store them to /var/log/fluent/access.%Y-%m-%d
    # Of course, you can control how you partition your data
    # with the time_slice_format option.
    <match myapp.access>
      @type file
      path /var/log/fluent/access
    </match>
`;
    expect(FluentBitSchema.isFluentBitConfiguration(fluentDConfig)).toBe(false);
  });
});
