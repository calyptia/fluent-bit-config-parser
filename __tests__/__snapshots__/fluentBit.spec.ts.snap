// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fluent Bit Utilities Retrieves the right tokens set from the given section. 1`] = `
Array [
  Object {
    "col": 1,
    "filePath": "<PROJECT_ROOT>/__fixtures__/basic/basic.conf",
    "line": 1,
    "lineBreaks": 0,
    "offset": 0,
    "text": "[",
    "toString": [Function],
    "type": "OPEN_BLOCK",
    "value": "[",
  },
  Object {
    "col": 2,
    "filePath": "<PROJECT_ROOT>/__fixtures__/basic/basic.conf",
    "line": 1,
    "lineBreaks": 0,
    "offset": 1,
    "text": "INPUT",
    "toString": [Function],
    "type": "INPUT",
    "value": "INPUT",
  },
  Object {
    "col": 7,
    "filePath": "<PROJECT_ROOT>/__fixtures__/basic/basic.conf",
    "line": 1,
    "lineBreaks": 0,
    "offset": 6,
    "text": "]",
    "toString": [Function],
    "type": "CLOSE_BLOCK",
    "value": "]",
  },
  Object {
    "col": 5,
    "filePath": "<PROJECT_ROOT>/__fixtures__/basic/basic.conf",
    "line": 2,
    "lineBreaks": 0,
    "offset": 12,
    "text": "Name        tail # some comment",
    "toString": [Function],
    "type": "PROPERTIES",
    "value": "Name tail # some comment",
  },
  Object {
    "col": 5,
    "filePath": "<PROJECT_ROOT>/__fixtures__/basic/basic.conf",
    "line": 3,
    "lineBreaks": 0,
    "offset": 48,
    "text": "Tag         tail.01",
    "toString": [Function],
    "type": "PROPERTIES",
    "value": "Tag tail.01",
  },
  Object {
    "col": 5,
    "filePath": "<PROJECT_ROOT>/__fixtures__/basic/basic.conf",
    "line": 4,
    "lineBreaks": 0,
    "offset": 72,
    "text": "Path        /var/log/system.log",
    "toString": [Function],
    "type": "PROPERTIES",
    "value": "Path /var/log/system.log",
  },
  Object {
    "col": 1,
    "filePath": "<PROJECT_ROOT>/__fixtures__/basic/basic.conf",
    "line": 6,
    "lineBreaks": 0,
    "offset": 105,
    "text": "[",
    "toString": [Function],
    "type": "OPEN_BLOCK",
    "value": "[",
  },
  Object {
    "col": 2,
    "filePath": "<PROJECT_ROOT>/__fixtures__/basic/basic.conf",
    "line": 6,
    "lineBreaks": 0,
    "offset": 106,
    "text": "OUTPUT",
    "toString": [Function],
    "type": "OUTPUT",
    "value": "OUTPUT",
  },
  Object {
    "col": 8,
    "filePath": "<PROJECT_ROOT>/__fixtures__/basic/basic.conf",
    "line": 6,
    "lineBreaks": 0,
    "offset": 112,
    "text": "]",
    "toString": [Function],
    "type": "CLOSE_BLOCK",
    "value": "]",
  },
  Object {
    "col": 5,
    "filePath": "<PROJECT_ROOT>/__fixtures__/basic/basic.conf",
    "line": 7,
    "lineBreaks": 0,
    "offset": 118,
    "text": "Name s3",
    "toString": [Function],
    "type": "PROPERTIES",
    "value": "Name s3",
  },
  Object {
    "col": 5,
    "filePath": "<PROJECT_ROOT>/__fixtures__/basic/basic.conf",
    "line": 8,
    "lineBreaks": 0,
    "offset": 130,
    "text": "Match *",
    "toString": [Function],
    "type": "PROPERTIES",
    "value": "Match *",
  },
  Object {
    "col": 5,
    "filePath": "<PROJECT_ROOT>/__fixtures__/basic/basic.conf",
    "line": 9,
    "lineBreaks": 0,
    "offset": 142,
    "text": "bucket your-bucket",
    "toString": [Function],
    "type": "PROPERTIES",
    "value": "bucket your-bucket",
  },
  Object {
    "col": 5,
    "filePath": "<PROJECT_ROOT>/__fixtures__/basic/basic.conf",
    "line": 10,
    "lineBreaks": 0,
    "offset": 165,
    "text": "region us-east-1",
    "toString": [Function],
    "type": "PROPERTIES",
    "value": "region us-east-1",
  },
  Object {
    "col": 5,
    "filePath": "<PROJECT_ROOT>/__fixtures__/basic/basic.conf",
    "line": 11,
    "lineBreaks": 0,
    "offset": 186,
    "text": "store_dir /home/ec2-user/buffer",
    "toString": [Function],
    "type": "PROPERTIES",
    "value": "store_dir /home/ec2-user/buffer",
  },
  Object {
    "col": 5,
    "filePath": "<PROJECT_ROOT>/__fixtures__/basic/basic.conf",
    "line": 12,
    "lineBreaks": 0,
    "offset": 222,
    "text": "total_file_size 50M",
    "toString": [Function],
    "type": "PROPERTIES",
    "value": "total_file_size 50M",
  },
  Object {
    "col": 5,
    "filePath": "<PROJECT_ROOT>/__fixtures__/basic/basic.conf",
    "line": 13,
    "lineBreaks": 0,
    "offset": 246,
    "text": "upload_timeout 10m",
    "toString": [Function],
    "type": "PROPERTIES",
    "value": "upload_timeout 10m",
  },
  Object {
    "col": 1,
    "filePath": "<PROJECT_ROOT>/__fixtures__/basic/basic.conf",
    "line": 15,
    "lineBreaks": 0,
    "offset": 266,
    "text": "[",
    "toString": [Function],
    "type": "OPEN_BLOCK",
    "value": "[",
  },
  Object {
    "col": 2,
    "filePath": "<PROJECT_ROOT>/__fixtures__/basic/basic.conf",
    "line": 15,
    "lineBreaks": 0,
    "offset": 267,
    "text": "OUTPUT",
    "toString": [Function],
    "type": "OUTPUT",
    "value": "OUTPUT",
  },
  Object {
    "col": 8,
    "filePath": "<PROJECT_ROOT>/__fixtures__/basic/basic.conf",
    "line": 15,
    "lineBreaks": 0,
    "offset": 273,
    "text": "]",
    "toString": [Function],
    "type": "CLOSE_BLOCK",
    "value": "]",
  },
  Object {
    "col": 5,
    "filePath": "<PROJECT_ROOT>/__fixtures__/basic/basic.conf",
    "line": 16,
    "lineBreaks": 0,
    "offset": 279,
    "text": "Name        splunk",
    "toString": [Function],
    "type": "PROPERTIES",
    "value": "Name splunk",
  },
  Object {
    "col": 5,
    "filePath": "<PROJECT_ROOT>/__fixtures__/basic/basic.conf",
    "line": 17,
    "lineBreaks": 0,
    "offset": 302,
    "text": "Match       *",
    "toString": [Function],
    "type": "PROPERTIES",
    "value": "Match *",
  },
  Object {
    "col": 5,
    "filePath": "<PROJECT_ROOT>/__fixtures__/basic/basic.conf",
    "line": 18,
    "lineBreaks": 0,
    "offset": 320,
    "text": "Host        127.0.0.1",
    "toString": [Function],
    "type": "PROPERTIES",
    "value": "Host 127.0.0.1",
  },
  Object {
    "col": 5,
    "filePath": "<PROJECT_ROOT>/__fixtures__/basic/basic.conf",
    "line": 19,
    "lineBreaks": 0,
    "offset": 346,
    "text": "Port        8088",
    "toString": [Function],
    "type": "PROPERTIES",
    "value": "Port 8088",
  },
  Object {
    "col": 5,
    "filePath": "<PROJECT_ROOT>/__fixtures__/basic/basic.conf",
    "line": 20,
    "lineBreaks": 0,
    "offset": 367,
    "text": "TLS         On",
    "toString": [Function],
    "type": "PROPERTIES",
    "value": "TLS On",
  },
  Object {
    "col": 5,
    "filePath": "<PROJECT_ROOT>/__fixtures__/basic/basic.conf",
    "line": 21,
    "lineBreaks": 0,
    "offset": 386,
    "text": "TLS.Verify  Off",
    "toString": [Function],
    "type": "PROPERTIES",
    "value": "TLS.Verify Off",
  },
  Object {
    "col": 5,
    "filePath": "<PROJECT_ROOT>/__fixtures__/basic/basic.conf",
    "line": 22,
    "lineBreaks": 0,
    "offset": 406,
    "text": "Message_Key my_key",
    "toString": [Function],
    "type": "PROPERTIES",
    "value": "Message_Key my_key",
  },
  Object {
    "col": 5,
    "filePath": "<PROJECT_ROOT>/__fixtures__/basic/basic.conf",
    "line": 23,
    "lineBreaks": 0,
    "offset": 429,
    "text": "add_label pipeline_id a21fd551-095b-4271-acf0-c2fdb3161b84",
    "toString": [Function],
    "type": "PROPERTIES",
    "value": "add_label pipeline_id a21fd551-095b-4271-acf0-c2fdb3161b84",
  },
]
`;

exports[`Fluent Bit Utilities Simpler section to tokens. 1`] = `
Array [
  Object {
    "col": 5,
    "filePath": "/nowhere/ephemeral.conf",
    "line": 2,
    "lineBreaks": 0,
    "offset": 5,
    "text": "[",
    "toString": [Function],
    "type": "OPEN_BLOCK",
    "value": "[",
  },
  Object {
    "col": 6,
    "filePath": "/nowhere/ephemeral.conf",
    "line": 2,
    "lineBreaks": 0,
    "offset": 6,
    "text": "INPUT",
    "toString": [Function],
    "type": "INPUT",
    "value": "INPUT",
  },
  Object {
    "col": 11,
    "filePath": "/nowhere/ephemeral.conf",
    "line": 2,
    "lineBreaks": 0,
    "offset": 11,
    "text": "]",
    "toString": [Function],
    "type": "CLOSE_BLOCK",
    "value": "]",
  },
  Object {
    "col": 7,
    "filePath": "/nowhere/ephemeral.conf",
    "line": 3,
    "lineBreaks": 0,
    "offset": 19,
    "text": "Name        tail # some comment",
    "toString": [Function],
    "type": "PROPERTIES",
    "value": "Name tail # some comment",
  },
  Object {
    "col": 7,
    "filePath": "/nowhere/ephemeral.conf",
    "line": 4,
    "lineBreaks": 0,
    "offset": 57,
    "text": "Tag         tail.01",
    "toString": [Function],
    "type": "PROPERTIES",
    "value": "Tag tail.01",
  },
  Object {
    "col": 7,
    "filePath": "/nowhere/ephemeral.conf",
    "line": 5,
    "lineBreaks": 0,
    "offset": 83,
    "text": "Path        /var/log/system.log",
    "toString": [Function],
    "type": "PROPERTIES",
    "value": "Path /var/log/system.log",
  },
]
`;
